### PR TITLE
update gardener version

### PR DESCRIPTION
**Pull request by deitch**
Signed-off-by: Avi Deitcher <avi@deitcher.net>

**How to categorize this PR?**

/area open-source
/kind cleanup

**What this PR does / why we need it**:

* revendor on the latest gardener, which required some significant changes
* document how to run locally and how to contribute

**Special notes for your reviewer**:

Depends on [Gardener PR 4383](http://github.com/gardener/gardener/pull/4383) to work fully correctly, so don't merge in until that happens

cc @rfranzke 

**Release note**:

```other developer
Update latest gardener base and document how to work with it
```

### CODE REVIEW

**Comment by gardener-robot-ci-1** [comment]
Thank you @deitch for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

**Comment by deitch** [comment]
Oh drat. The conflicts on this are enormous.

**Comment by deitch** [comment]
What a mess. But I think I got it.

**Comment by deitch** [comment]
Now that [Gardener PR 4383](https://github.com/gardener/gardener/pull/4383) is in, I updated `go.mod` to the latest commit of Gardener, then `make revendor` and `make generate`. Committed changes and pushed out, so this is up to date. Running `make check` in parallel.

**Comment by deitch** [comment]
All errors from `make check` cleared, reran `make generate`. @rfranzke I think this is ready for you to take a look.

File: Makefile [hunk 1]
```diff
@@ -48,6 +49,21 @@ start:
 		--leader-election=$(LEADER_ELECTION) \
 		--webhook-config-server-host=0.0.0.0 \
 		--webhook-config-server-port=$(WEBHOOK_CONFIG_PORT) \
+    --webhook-config-service-port=$(WEBHOOK_CONFIG_SERVICE_PORT) \
```
Comment (rfranzke) [inline Makefile:52]: Nit: inconsistent indentation

File: Makefile [hunk 2]
```diff
@@ -48,6 +49,21 @@ start:
...
+	@LEADER_ELECTION_NAMESPACE=garden GO111MODULE=on GARDENER_SHOOT_CLIENT=external dlv debug \
+		--build-flags="-mod=vendor" \
+		./cmd/$(EXTENSION_PREFIX)-$(NAME) \
+		-- \
+		--config-file=./example/00-componentconfig.yaml \
+		--ignore-operation-annotation=$(IGNORE_OPERATION_ANNOTATION) \
+		--leader-election=$(LEADER_ELECTION) \
+		--webhook-config-server-host=0.0.0.0 \
+		--webhook-config-server-port=$(WEBHOOK_CONFIG_PORT) \
+    --webhook-config-service-port=$(WEBHOOK_CONFIG_SERVICE_PORT) \
```
Comment (rfranzke) [inline Makefile:66]: Nit: inconsistent indentation

File: Makefile [hunk 3]
```diff
@@ -26,6 +26,7 @@ IGNORE_OPERATION_ANNOTATION := true
 WEBHOOK_CONFIG_PORT	:= 8443
 WEBHOOK_CONFIG_MODE	:= url
 WEBHOOK_CONFIG_URL	:= host.docker.internal:$(WEBHOOK_CONFIG_PORT)
+WEBHOOK_CONFIG_SERVICE_PORT	:= 443
```
Comment (rfranzke) [inline Makefile:29]: Nit: inconsistent indentation

File: pkg/apis/packet/types_controlplane.go [hunk 4]
```diff
@@ -23,4 +23,11 @@ import (
 // ControlPlaneConfig contains configuration settings for the control plane.
 type ControlPlaneConfig struct {
 	metav1.TypeMeta
+
+	Persistence *Persistence `json:"persistence,omitempty"`
+}
+
+// Persistence contains configuration options for shoot persistent storage.
+type Persistence struct {
+	Enabled *bool `json:"enabled,omitempty"`
```
Comment (rfranzke) [inline pkg/apis/packet/types_controlplane.go:32]: I guess those things are not related to this PR and should be removed, or? (same for the changes in the `v1alpha1` package)

**Comment by PavelPavlov46** [comment]
hey @deitch , I see "Files changed 1,843". Could it be that something needs to be added to .gitignore or something?

File: pkg/controller/controlplane/valuesprovider.go [hunk 5]
```diff
@@ -192,10 +190,32 @@ func getControlPlaneChartValues(
...
+	cp *extensionsv1alpha1.ControlPlane,
 	cluster *extensionscontroller.Cluster,
 	credentials *packet.Credentials,
 ) (map[string]interface{}, error) {
 
-	return map[string]interface{}{}, nil
+	_, err := vp.decodeControlPlaneConfig(cp)
+	if err != nil {
+		return nil, errors.Wrapf(err, "decoding control plane config")
+	}
```
Comment (rfranzke) [inline pkg/controller/controlplane/valuesprovider.go:202]: Why was this added? It seems to not have any effect?

**Comment by deitch** [comment]
I am not sure @PavelPavlov46; I did a `make revendor` which pulls in every change in latest gardener and all of their dependencies.

File: pkg/controller/worker/machines.go [hunk 6]
```diff
@@ -138,6 +138,10 @@ func (w *workerDelegate) generateMachineConfig(ctx context.Context) error {
 			"secret": map[string]interface{}{
 				"cloudConfig": string(pool.UserData),
 			},
+			"credentialsSecretRef": map[string]interface{}{
+				"name":      "cloudprovider",
```
Comment (rfranzke) [inline pkg/controller/worker/machines.go:142]: You could use `v1beta1constants.SecretNameCloudProvider`

File: pkg/controller/worker/machine_dependencies.go [hunk 7]
```diff
@@ -14,12 +14,281 @@
...
+	if err := shoot.List(ctx, shootNodes); err != nil {
+		return fmt.Errorf("failed to get shoot nodes: %v", err)
+	}
+	// go through each node, for each one without the right annotation, get the private network
+	var targetCidrs []string
+	for _, n := range shootNodes.Items {
+		if n.Annotations == nil || n.Annotations[equinixMetalPrivateNetworkAnnotations] == "" {
+			// we didn't have it, so get it from the Equinix Metal API, and save it
+			deviceID, err := deviceIDFromProviderID(n.Spec.ProviderID)
+			fmt.Printf("deviceID %s err %#v", deviceID, err)
```
Comment (rfranzke) [inline pkg/controller/worker/machine_dependencies.go:70]: Is this a left-over?

File: pkg/controller/worker/machine_dependencies.go [hunk 8]
```diff
@@ -14,12 +14,281 @@
...
+	return list
+}
+
+// getNodePrivateNetwork use the Equinix Metal API to get the CIDR of the private network given a providerID.
+func getNodePrivateNetwork(deviceID string, kClient client.Client, namespace string) (string, error) {
+	ctx := context.Background()
+	secretRef := corev1.SecretReference{
+		Name:      v1beta1constants.SecretNameCloudProvider,
+		Namespace: namespace,
+	}
```
Comment (rfranzke) [inline pkg/controller/worker/machine_dependencies.go:245]: Hm, we should extend the interface and also pass the `Worker` object to the `CleanupMachineDependencies` function. It's not nice to duplicate the secret ref here (only works by accident). Let's keep it this way for now, I'll put it on my TODO list to update the interface.

File: pkg/controller/worker/machine_dependencies.go [hunk 9]
```diff
@@ -14,12 +14,281 @@
...
+			keys[entry] = true
+			list = append(list, entry)
+		}
+	}
+	return list
+}
+
+// getNodePrivateNetwork use the Equinix Metal API to get the CIDR of the private network given a providerID.
+func getNodePrivateNetwork(deviceID string, kClient client.Client, namespace string) (string, error) {
+	ctx := context.Background()
```
Comment (rfranzke) [inline pkg/controller/worker/machine_dependencies.go:241]: Don't use `context.Background`, rather pass the `ctx` from `CleanupMachineDependencies`

File: pkg/controller/worker/machine_dependencies.go [hunk 10]
```diff
@@ -14,12 +14,281 @@
...
+	}
+	return c.Update(ctx, deploy)
+}
+
+func (w *workerDelegate) UpdateConfigMap(ctx context.Context) error {
+	ns := w.worker.GetNamespace()
+	c := w.Client()
+
+	// get the private IPs from the shoot nodes
+	_, shoot, err := util.NewClientForShoot(ctx, c, ns, client.Options{})
```
Comment (rfranzke) [inline pkg/controller/worker/machine_dependencies.go:152]:
```suggestion
      _, shootClient, err := util.NewClientForShoot(ctx, c, ns, client.Options{})
```

File: pkg/controller/worker/machine_dependencies.go [hunk 11]
```diff
@@ -14,12 +14,281 @@
...
+		return err
+	}
+
+	shootNodes := &corev1.NodeList{}
+	if err := shoot.List(ctx, shootNodes); err != nil {
+		return fmt.Errorf("failed to get shoot nodes: %v", err)
+	}
+
+	// go through each node, for each one without the right annotation, get the private network
+	var targetCidrs []string
```
Comment (rfranzke) [inline pkg/controller/worker/machine_dependencies.go:163]: You could use `sets.String` - it also allows you to delete the `unique` function and its `List()` method implicitly sorts

File: pkg/controller/worker/machine_dependencies.go [hunk 12]
```diff
@@ -14,12 +14,281 @@
...
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	vpnSeedConfigMap                      = "equinix-vpn-seed-config"
```
Comment (rfranzke) [inline pkg/controller/worker/machine_dependencies.go:36]:
```suggestion
      vpnSeedConfigMap = "gardener-extension-provider-equinixmetal-vpn-seed-config"
```

optional: (for consistency/to make clear what this is)

File: pkg/controller/worker/machine_dependencies.go [hunk 13]
```diff
@@ -14,12 +14,281 @@
...
+	vpnSeed                               = "vpn-seed"
+	apiServerDeploy                       = "kube-apiserver"
```
Comment (rfranzke) [inline pkg/controller/worker/machine_dependencies.go:38]: Remove this and use `v1beta1constants.DeploymentNameKubeAPIServer`

File: pkg/controller/worker/machine_dependencies.go [hunk 14]
```diff
@@ -14,12 +14,281 @@
...
+		ctrs[i] = ctr
+		changed = true
+	}
+	if !changed {
+		return nil
+	}
+	return c.Update(ctx, deploy)
+}
+
+func (w *workerDelegate) UpdateConfigMap(ctx context.Context) error {
```
Comment (rfranzke) [inline pkg/controller/worker/machine_dependencies.go:147]: What calls this function 👀 ? And what is the `ConfigMap` used for?

File: Makefile (hunk 1 above)
Comment (deitch) [reply Makefile:52]: Cleaned up, push coming

File: Makefile (hunk 2 above)
Comment (deitch) [reply Makefile:66]: Cleaned up, push coming

File: Makefile (hunk 3 above)
Comment (deitch) [reply Makefile:29]: Cleaned up, push coming

File: pkg/controller/controlplane/valuesprovider.go (hunk 5 above)
Comment (deitch) [reply pkg/controller/controlplane/valuesprovider.go:202]: It was related to storage. When I removed the storage, we didn't need the return value, but I forgot to delete the whole thing.

Removed, push on its way.

File: pkg/apis/packet/types_controlplane.go (hunk 4 above)
Comment (deitch) [reply pkg/apis/packet/types_controlplane.go:32]: Good point. Lingering from the storage work. Removed.

File: pkg/controller/worker/machines.go (hunk 6 above)
Comment (deitch) [reply pkg/controller/worker/machines.go:142]: I like that better.

File: pkg/controller/worker/machine_dependencies.go (hunk 7 above)
Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go:70]: Oops.

File: pkg/controller/worker/machine_dependencies.go (hunk 9 above)
Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go:241]: Done

File: pkg/controller/worker/machine_dependencies.go (hunk 10 above)
Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go:152]: Done

File: pkg/controller/worker/machine_dependencies.go (hunk 11 above)
Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go:163]: Which?

File: pkg/controller/worker/machine_dependencies.go (hunk 12 above)
Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go:36]: Done. Good idea.

File: pkg/controller/worker/machine_dependencies.go (hunk 13 above)
Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go:38]: Done

File: pkg/controller/worker/machine_dependencies.go (hunk 14 above)
Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go:147]: It is a legacy. I originally went down two paths for getting the vpn working, one with a configmap. I can remove it now.

Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go:147]: Done

File: pkg/controller/worker/machine_dependencies.go (hunk 11 above)
Comment (rfranzke) [reply pkg/controller/worker/machine_dependencies.go:163]: https://github.com/kubernetes/apimachinery/blob/57f2a0733447cfd41294477d833cce6580faaca3/pkg/util/sets/string.go

Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go:163]: Eh, I don't care too much. It works.

**Comment by deitch** [comment]
I just pushed it out, I think I addressed all of your comments @rfranzke

File: pkg/controller/worker/machine_dependencies.go (hunk 8 above)
Comment (rfranzke) [reply pkg/controller/worker/machine_dependencies.go:161]: I was confused about that one, the `Worker` object should already be stored at the `workerDelegate` struct. Hence, you could use `w.worker.Spec.SecretRef` instead of redefining this `SecretReference` here, right?

File: pkg/controller/worker/machine_dependencies.go (hunk 11 above)
Comment (rfranzke) [reply pkg/controller/worker/machine_dependencies.go]: Fair enough, just a suggestion. Resolving this.

File: pkg/controller/worker/machine_dependencies.go (hunk 8 above)
Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go:161]: I don't know. You tell me. If I can just pull it from there, then I can pass it in. Can I?

File: pkg/controller/worker/machine_dependencies.go [hunk 15]
```diff
@@ -14,12 +14,197 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	vpnSeed                               = "vpn-seed"
+	apiServerDeploy                       = v1beta1constants.DeploymentNameKubeAPIServer
+	nodeNetworkEnvVar                     = "NODE_NETWORK"
+	equinixMetalPrivateNetworkAnnotations = "metal.equinix.com/network/4/private"
```
Comment (rfranzke) [inline pkg/controller/worker/machine_dependencies.go:39]: How does this work/was this tested? When I try to annotate a `Node` with this key then the API server rejects it:

```
$ k annotate no ip-10-222-27-202.eu-west-1.compute.internal metal.equinix.com/network/4/private=foo
The Node "ip-10-222-27-202.eu-west-1.compute.internal" is invalid: metadata.annotations: Invalid value: "metal.equinix.com/network/4/private": a qualified name must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]') with an optional DNS subdomain prefix and '/' (e.g. 'example.com/MyName')
```

File: pkg/controller/worker/machine_dependencies.go [hunk 16]
```diff
@@ -14,12 +14,197 @@
...
 	return nil
 }
 
 func (w *workerDelegate) CleanupMachineDependencies(ctx context.Context) error {
-	return nil
+	ns := w.worker.GetNamespace()
+	c := w.Client()
+
+	// get the private IPs and providerIDs from the shoot nodes
+	_, shoot, err := util.NewClientForShoot(ctx, c, ns, client.Options{})
```
Comment (rfranzke) [inline pkg/controller/worker/machine_dependencies.go:53]:
```suggestion
	_, shootClient, err := util.NewClientForShoot(ctx, c, ns, client.Options{})
```

File: pkg/controller/worker/machine_dependencies.go (hunk 8 above)
Comment (rfranzke) [reply pkg/controller/worker/machine_dependencies.go:161]: Yes

File: pkg/controller/worker/machine_dependencies.go (hunk 15 above)
Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go]: Fixed. And a separate commit this time. And aligned with what CCM is using.

File: pkg/controller/worker/machine_dependencies.go (hunk 16 above)
Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go:53]: Accepted and in.

File: pkg/controller/worker/machine_dependencies.go (hunk 8 above)
Comment (deitch) [reply pkg/controller/worker/machine_dependencies.go]: Done

**Comment by deitch** [comment]
Addressed your fixes and suggestions @rfranzke . And the last bunch were in separate commits 😁 

Once everything is good, we should squash to a single one

File: pkg/controller/worker/machines.go [hunk 17]
```diff
@@ -138,6 +138,10 @@ func (w *workerDelegate) generateMachineConfig(ctx context.Context) error {
 			"secret": map[string]interface{}{
 				"cloudConfig": string(pool.UserData),
 			},
+			"credentialsSecretRef": map[string]interface{}{
+				"name":      v1beta1constants.SecretNameCloudProvider,
+				"namespace": w.worker.Namespace,
+			},
```
Comment (rfranzke) [inline pkg/controller/worker/machines.go:144]:
```suggestion
			"credentialsSecretRef": map[string]interface{}{
				"name":      w.worker.Spec.SecretRef.Name,
				"namespace": w.worker.Spec.SecretRef.Namespace,
			},
```

Comment (deitch) [reply pkg/controller/worker/machines.go:144]: Good one. Fixed.

**Comment by deitch** [comment]
haha! The CI is fixed!

**Review by rfranzke — APPROVED** [review]
/lgtm
